rfc-171-remove-legacy-browser-js-support.md

```diff
@@ -0,0 +1,128 @@
+---
+status: proposed
+implementation: proposed
+status_last_reviewed:
+---
+
+# Remove JavaScript support for legacy browsers
+
+## Summary
+
+This RFC proposes removing JavaScript (JS) support for old (legacy) browsers such as Internet Explorer across the GOV.UK website. This is because:
+
+- the browsers in question are no longer supported by their vendors
+- we cannot offer these browsers an entirely equal experience because `govuk-frontend` has already dropped JS support for legacy browsers
+- the percentage of users likely to be affected is well below 1%, and will not increase
+- we build progressively and the site will remain legible, navigable and usable for all users regardless of this change
+
+The benefits of this are:
+
+- reduced burden on developers to write extra code to support legacy browsers
+- minor performance improvements for other users as we stop branching code and including polyfills
+- we can embrace modern JS techniques, which could improve and simplify our code
+
+The drawbacks are:
+
+- users of legacy browsers will no longer be included in our analytics tracking
+
+## Problem
+
+This RFC is taking a broad approach to legacy browsers, but the main point of contention is likely to be Internet Explorer 11, the legacy browser with the most significant number of ongoing visitors.
+
+At the time of writing the service manual (which was last updated in 2022) says "You will still need to support Internet Explorer 11 if your analytics data shows at least 2% of your users arriving at the service's start page are using it." It [recommends testing](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices) only in Edge of all the Microsoft browsers.
+
+The Design System team have updated `govuk-frontend` to use [ES modules in version 5](https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/#javascript), released in 2023. [ES modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) are not recognised by legacy browsers, notably Internet Explorer 11 and below. These browsers are no longer supported by Microsoft. We need to keep up to date with `govuk-frontend`, which means deciding how we approach JS support for IE11. The Design System has their own definitions for [browser support](https://frontend.design-system.service.gov.uk/browser-support/).
+
+The presence of any modern JS such as ES modules will cause errors for older browsers. This is because they will attempt to parse all JS before running it, encounter unfamiliar syntax, and error. No JS will run on older browsers once any modern JS is introduced, which means that any download of JS will be wasted.
+
+## Proposal
```

Comment:
A few other questions I had that didn't have a clear place to comment are:

- is a team committing to this piece of work as part of their OKRs or is it something done on the side of team work? (just to gauge resources we're putting into it)
- I assume a consequence is potentially breaking changes to govuk_publishing_components, do we have any idea on the impact of that for other teams and what support might be available? will things happen all at once or gradually?
- This will presumably mean decoupling from [Rails asset_pipeline](https://github.com/alphagov/govuk_publishing_components/issues/505) for JS (yay) do we have a proposal for what replaces it?

Reply:
This is a web 2 OKR, getting consensus for the direction is step 1 on the plan. The details of next steps need some thought and planning and will be hammered out by frontenders across GOV.UK

I would expect the breaking changes to be minimal but would expect a similar model to the dart sass work: - the team handles the website apps, writes up what others will need to do, and remains available for pairing if something gets tricky.